src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -70,22 +72,31 @@ public JsonClassInfo ElementClassInfo
 
         public void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Check if we are trying to build the sorted cache.
-            if (frame.PropertyRefCache == null)
-            {
-                return;
-            }
+            Debug.Assert(frame.PropertyRefCache != null);
 
-            List<PropertyRef> newList;
-            if (_propertyRefsSorted != null)
-            {
-                newList = new List<PropertyRef>(_propertyRefsSorted);
-                newList.AddRange(frame.PropertyRefCache);
-                _propertyRefsSorted = newList.ToArray();
-            }
-            else
+            lock (_cacheLock)
```

Comment:
Can you describe what the race condition was and how this addresses it?

Also, what are the access patterns to UpdateSortedPropertyCache?  How often is it called?

Reply:
I added detail to the PR description. UpdateSortedPropertyCache is only called when a given object is done being deserialized, and there are "new" properties that need to be appended - now explained by the PR description.